File: 2025/golang/day1/part1.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
)

func solve(input string) {
	file, err := os.Open(input)
	if err != nil {
		panic(err)
	}

	defer file.Close()

	zeroCount := 0
	position := 50
	scanner := bufio.NewScanner(file)
	direction := 1
	for scanner.Scan() {
		line := scanner.Text()
		if first := line[0]; first == 'L' {
			direction = -1
		} else {
			direction = 1
		}

		steps, err := strconv.Atoi(line[1:])
		if err != nil {
			panic(err)
		}

		position = rotate(position, direction, steps)
		if position == 0 {
			zeroCount++
		}
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	fmt.Println("Total:", zeroCount)
}

func rotate(position int, direction int, steps int) int {
	position = (position + direction*steps) % 100
	if position < 0 {
		position = 100 + position
	}

	return position
}

func main() {
	// input := "input_test.txt" // Expected: 3
	input := "input.txt" // 1055
	solve(input)
}

File: 2025/python/day1/part1.py
def main(file: str) -> None:
    zero_count = 0
    position = 50
    with open(file, "r") as f:
        for line in f.readlines():
            direction = -1 if line[0] == "L" else 1
            steps = int(line[1:])

            position = rotate(position, direction, steps)
            if position == 0:
                zero_count += 1

    print(zero_count)


def rotate(position: int, direction: int, steps: int) -> int:
    new_position = (position + direction * steps) % 100
    if new_position < 0:
        new_position = 100 - new_position

    return new_position


if __name__ == "__main__":
    input_test_str = "input_test.txt"  # Expected output: 3
    input_str = "input.txt"  # 1055
    main(input_str)
